File: src/102-binary-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val ?? 0;
    this.left = left ?? null;
    this.right = right ?? null;
  }
}

function buildTree(values: Array<number | null>): TreeNode | null {
  if (values.length === 0 || values[0] === null) return null

  const root = new TreeNode(values[0])
  const queue: TreeNode[] = [root]
  let i = 1;

  while (i < values.length && queue.length > 0) {
    const node = queue.shift()!;

    const leftValue = values[i];
    i++;

    if (leftValue !== undefined && leftValue !== null) {
      node.left = new TreeNode(leftValue)
      queue.push(node.left)
    }

    const rightValue = values[i];

    i++;
    if (rightValue !== undefined && rightValue !== null) {
      node.right = new TreeNode(rightValue)
      queue.push(node.right)
    }
  }

  return root
}


function levelOrder(root: TreeNode | null): number[][] {
  if (root === null || root.val === null) return []

  const result: number[][] = []
  let queue = [root]


  while (queue.length > 0) {
    const level: number[] = []
    const levelQueue = [...queue]
    queue = []

    for (const node of levelQueue) {
      if (node.val !== undefined && node.val !== null) level.push(node.val)

      if (node.left?.val !== undefined && node.left.val !== null) queue.push(node.left)
      if (node.right?.val !== undefined && node.right.val !== null) queue.push(node.right)
    }
    result.push(level)

  }

  return result
}

const input: (number | null)[] = [3, 9, 20, null, null, 15, 7];


const root = buildTree(input);
console.log(root)

console.log(levelOrder(root)); // [[3], [9, 20], [15, 7]]/ const: [[3],[9,20],[15,7]]

File: src_py/102-binary-tree.py
"""
102. Binary Tree Level Order Traversal
Difficulty: Medium

Given the root of a binary tree, return the level order traversal of its
nodes' values. (i.e., from left to right, level by level).
"""

from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: "Optional[TreeNode]" = None,
        right: "Optional[TreeNode]" = None,
    ):
        self.val = val
        self.left = left
        self.right = right


def build_tree_new(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 0
    while i < len(values) and queue:
        node = queue.popleft()
        if i < len(values):
            current = values[i]
            if current is not None:
                node.left = TreeNode(current)
                queue.append(node.left)
        i += 1
        if i < len(values):
            current = values[i]
            if current is not None:
                node.right = TreeNode(current)
                queue.append(node.right)
        i += 1
    return root


def build_tree(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values or values[0] is None:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1

    while i < len(values) and queue:
        node = queue.popleft()

        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])  # pyright: ignore
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])  # pyright: ignore
            queue.append(node.right)
        i += 1

    return root


def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    if root is None:
        return []

    result: List[List[int]] = []
    queue = deque([root])

    while queue:
        level: List[int] = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.append(level)

    return result


if __name__ == "__main__":
    node_list = [3, 9, 20, None, None, 15, 7]
    root = build_tree(node_list)
    new_root = build_tree_new(node_list)
    print("root node", new_root)
    assert level_order(root) == [[3], [9, 20], [15, 7]]
    assert level_order(build_tree([1])) == [[1]]
    assert level_order(build_tree([])) == []
    print("102 OK:", level_order(root))
